src/op_handlers/aux_heap_read.rs

```diff
@@ -2,24 +2,29 @@ use u256::U256;
 use zkevm_opcode_defs::MAX_OFFSET_TO_DEREF_LOW_U32;
 
 use crate::address_operands::address_operands_read;
+use crate::eravm_error::EraVmError;
 use crate::value::TaggedValue;
 use crate::{opcode::Opcode, state::VMState};
 
-pub fn _aux_heap_read(vm: &mut VMState, opcode: &Opcode) {
-    let (src0, _) = address_operands_read(vm, opcode);
+pub fn _aux_heap_read(vm: &mut VMState, opcode: &Opcode) -> Result<(), EraVmError> {
+    let (src0, _) = address_operands_read(vm, opcode)?;
     if src0.is_pointer {
-        panic!("Invalid operands for heap_read");
+        return Err(EraVmError::OperandError(
+            "Invalid operands for heap_read".to_string(),
+        ));
     }
 
     if src0.value > U256::from(MAX_OFFSET_TO_DEREF_LOW_U32) {
-        panic!("Address too large for heap_read");
+        return Err(EraVmError::OperandError(
+            "Address too large for heap_read".to_string(),
+        ));
     }
     let addr = src0.value.low_u32();
```

Comment:
It makes a ton more sense that `low_u32()` returns a `Result`. Then, it can have a proper error variant for out of bounds.

Reply:
Or, rather, have an `impl TryFrom<U256> for u32`.

Reply:
Adding `TryFrom` would be nice, but `U256` is an external type, so it's not possible to implement it.

Same goes for the method `low_u32()`, it's not possible to modify it.

I can't think of any other way of doing this than what is currently done (manual check, followed by assignment to variable).

Reply:
Sad. We could use a function or a local trait that mimics `TryFrom`'s behavior. If the trait is local to our crate then the orphan rule is satisfied.

Reply:
Following this, `U256` actually implements `try_into()` `u32`, but it doesn't fit our case, because we are not trying to check wheter the value is lower or equal than `u32::MAX`, but rather `MAX_OFFSET_TO_DEREF_LOW_U32`, which is smaller.

So we could implement a trait like this:
```rs
pub trait IntoU32Address {
    fn into_u32_address(self, opcode: Variant) -> Result<u32, OperandError>;
}

impl IntoU32Address for U256 {
    fn into_u32_address(self, opcode: Variant) -> Result<u32, OperandError> {
        if self > U256::from(MAX_OFFSET_TO_DEREF_LOW_U32) {
            return Err(OperandError::InvalidSrcAddress(opcode));
        }
        Ok(self.low_u32())
    }
}
```

And use it like this:

```rs
pub fn aux_heap_read(vm: &mut VMState, opcode: &Opcode) -> Result<(), EraVmError> {
    let (src0, _) = address_operands_read(vm, opcode)?;
    if src0.is_pointer {
        return Err(OperandError::InvalidSrcPointer(opcode.variant).into());
    }

    let addr = src0.value.into_u32_address(opcode.variant)?; // <-- HERE
    let gas_cost = vm.current_frame_mut()?.aux_heap.expand_memory(addr + 32);
    vm.current_frame_mut()?.gas_left -= gas_cost;

    let value = vm.current_frame_mut()?.aux_heap.read(addr);
    vm.set_register(opcode.dst0_index, TaggedValue::new_raw_integer(value));

    if opcode.alters_vm_flags {
        // This flag is set if .inc is present
        vm.set_register(
            opcode.dst1_index,
            TaggedValue::new_raw_integer(U256::from(addr + 32)),
        );
    }
    Ok(())
}
```